### PR TITLE
Handle test mappings between V1's move_check/unit_test/*.unit_text.exp and V2's unit_test/test/*.exp

### LINKED ISSUE #12810: [Bug][move-compiler-v2] PR 12780 added test files which aren't handled by update_v1_diff.sh script
# 🐛 Bug

Need to fix the script to handle them.

### PR DESCRIPTION
**Pull request by brmataptos**
## Description

Fix #12810 by adding mappings for unit_test options, and not filtering attribute-related tests.

Note that V1 has a complicated scheme to run test files in `/unit_test/` directory in both "test mode" and "non-test" mode,
- in the unit_test dir, if in addition to `foo.move` there is a `foo.unit_test`, then 
  - the ordinary test is run with `foo.move` in "non-test mode" to yield output in `foo.exp`
  - the same test input `foo.move` is run in "test mode" to yield output in `foo.unit_test.exp`

Meanwhile, the `testdiff` script (core of `update_v1_diff.sh`) which tries to map test directories from v2 to v1, doesn't support name changes, but has a `UNIT_PATH_REMAP` mapping v2 test dir name to v1 test dir name, and applies it to all expected `.exp` files to add to `v1.matched` files. (*)

This PR adds a pattern-match for remaps so that a `UNIT_PATH_REMAP` entry mapping  containing a dot, `("v2-dir", "v1-dir.ext")`, maps `v2-dir/file.exp` to `v1-dir/file.ext.exp`.  The entry `("unit_test/test", "unit_test.unit_test")` is used to match v2 file `unit_test/test/foo.exp` to v1 file `unit_test/foo.unit_test.exp`.

Note that I left some DEBUG-guarded `eprintln!` statements since I suspect this may need further editing in the future.

(*) Technically, it starts with `.move` files and looks for corresponding `.exp` files.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [x] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)
   move-compiler-v2
## How Has This Been Tested?
Ran the update_v1_diff.sh scripts locally, and included outputs in this PR.

## Key Areas to Review
Whether new tests in PR might be bad, other than those already mentioned in Issue 12789.

## Checklist
- [X] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I identified and added all stakeholders and component owners affected by this change as reviewers
- [X] I tested both happy and unhappy path of the functionality
- [X] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by rahxephon89 — APPROVED** [review]
It would be nice if you could give a background of the issue (how the unit_test folder and extension works in V2) and why this fix is necessary in the PR description. Thanks

File: third_party/move/move-compiler-v2/tools/testdiff/src/main.rs [hunk 1]
```diff
@@ -53,10 +56,7 @@ const V2_E2E_ROOT: &str = "move-compiler-v2/transactional-tests/tests";
 /// Ignored directories in the move_check tree.
 static IGNORED_UNIT_PATHS: Lazy<BTreeSet<String>> = Lazy::new(|| {
     let mut set = BTreeSet::new();
-    set.insert("parser".to_string());
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tools/testdiff/src/main.rs:56]: Presumably, you needed to not ignore this because there were some parser tests you wanted to match against for the attributes stuff?

This change means that we would either (a) need to port over unmatched parser tests into v2, even though we currently use the exact same parser, (b) be okay with unmatched parser tests.

File: third_party/move/move-compiler-v2/tools/testdiff/src/main.rs [hunk 2]
```diff
@@ -201,8 +187,11 @@ fn collect_tests(root: &str) -> anyhow::Result<BTreeSet<PathBuf>> {
     let mut result = BTreeSet::new();
     for entry in WalkDir::new(root) {
         let path = entry?.into_path();
-        if path.display().to_string().ends_with(".move") {
+        let path_str = path.display().to_string();
+        if path_str.ends_with(".move") {
             result.insert(path);
+        } else if path_str.ends_with(".unit_test") {
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tools/testdiff/src/main.rs:193]: Maybe worth documenting this with a comment (like you have in the PR description)?

File: third_party/move/move-compiler-v2/tools/testdiff/src/main.rs [hunk 3]
```diff
@@ -214,10 +203,14 @@ fn classify_tests(
     for_v2: bool,
 ) -> BTreeMap<TestKey, PathBuf> {
     let mut result = BTreeMap::new();
+    let split_match = Regex::new("^([^.]*)\\.([^.]*)$").unwrap();
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tools/testdiff/src/main.rs:206]: It would be preferable to use standard library methods like `extension` (https://doc.rust-lang.org/std/path/struct.Path.html#method.extension) and `file_prefix` here.

File: third_party/move/move-compiler-v2/tools/testdiff/src/main.rs [hunk 4]
```diff
@@ -231,8 +224,39 @@ fn classify_tests(
...
+                            "Parent {} matches, new_parent={}, file_with_ext={}",
+                            parent_str, new_parent, file_with_ext_move
+                        );
+                    }
+                    (new_parent, file_with_ext_move.to_string())
+                } else {
+                    (parent_str, file_name.to_string_lossy().to_string())
+                };
+            let key = TestKey::new(parent_str, file_name_str);
+            if DEBUG {
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tools/testdiff/src/main.rs:252]: For consisency, could we do what `warn()` does below, and have something like `debug()`, hiding the `if DEBUG` in various places (minimize some code dup, obscures reading of the code a bit)?

File: third_party/move/move-compiler-v2/tools/testdiff/src/main.rs [hunk 5]
```diff
@@ -7,13 +7,16 @@
...
 use once_cell::sync::Lazy;
+use regex::{self, Regex};
 use std::{
     collections::{BTreeMap, BTreeSet},
     fmt::{Display, Formatter},
     path::PathBuf,
 };
 use walkdir::WalkDir;
 
+const DEBUG: bool = false;
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tools/testdiff/src/main.rs:18]: Maybe you can say here that the expectation is to set this manually to true, when trying to debug this tool?

File: third_party/move/move-compiler-v2/tools/testdiff/src/main.rs (hunk 2 above)
Comment (vineethk) [reply third_party/move/move-compiler-v2/tools/testdiff/src/main.rs:193]: On the other hand, I see that there is indeed some documentation below, so this is like not needed.

File: Cargo.lock [hunk 6]
```diff
@@ -8180,7 +8180,7 @@ dependencies = [
  "aho-corasick",
  "bstr 1.9.0",
  "log",
- "regex-automata 0.4.3",
+ "regex-automata 0.4.6",
```
Comment (vineethk) [inline Cargo.lock:8183]: Do we need all these updates? Especially if end up using standard methods on paths?

File: Cargo.lock [hunk 7]
```diff
@@ -8777,7 +8777,7 @@ dependencies = [
  "httpdate",
  "itoa",
  "pin-project-lite",
- "socket2 0.4.10",
+ "socket2 0.5.5",
```
Comment (vineethk) [inline Cargo.lock:8780]: Another seemingly unrelated change. Reconsider updating these dependencies.

**Review by vineethk — COMMENTED** [review]
Looking good.

File: third_party/move/move-compiler-v2/tools/testdiff/src/main.rs (hunk 1 above)
Comment (brmataptos) [reply third_party/move/move-compiler-v2/tools/testdiff/src/main.rs:56]: At some point we need to worry about whether we are invoking the parser correctly in compiler v2.  For this reason, it's not a good idea to ignore the parser tests forever.

For example, are we recognizing V2 language extensions?  Test annotations?  etc.

File: third_party/move/move-compiler-v2/tools/testdiff/src/main.rs (hunk 3 above)
Comment (brmataptos) [reply third_party/move/move-compiler-v2/tools/testdiff/src/main.rs:206]: I found the file extension methods to be not as reliable as something better-defined like a Regexp.  (Which I guess is actually Regex, so maybe not so well-defined.)

File: third_party/move/move-compiler-v2/tools/testdiff/src/main.rs (hunk 4 above)
Comment (brmataptos) [reply third_party/move/move-compiler-v2/tools/testdiff/src/main.rs:252]: Annoyingly, unless we define a macro or have an `if` we can't avoid evaluating the parameters each time.  I don't feel ready to start creating macros in ordinary Rust code.  See the V1 compiler code.

File: third_party/move/move-compiler-v2/tools/testdiff/src/main.rs (hunk 1 above)
Comment (vineethk) [reply third_party/move/move-compiler-v2/tools/testdiff/src/main.rs:56]: Sure, but the same logic applies to expansion as well, which we currently ignore ...

It would be helpful for the reviewer when there is such a large change, the PR description mentions why it is being done (parser tests need to unignored here because ..., we will port parser tests as follow up, etc). I agree that "at some point we need to worry ...", my question was just to understand why we are doing it here.

File: third_party/move/move-compiler-v2/tools/testdiff/src/main.rs (hunk 3 above)
Comment (vineethk) [reply third_party/move/move-compiler-v2/tools/testdiff/src/main.rs:206]: In what way are they not as reliable? Those methods are well documented for cases like `blah.exp1.exp2` etc, whereas for a regex, I need to hunt a bit more to analyze the behavior.

File: third_party/move/move-compiler-v2/tools/testdiff/src/main.rs (hunk 4 above)
Comment (vineethk) [reply third_party/move/move-compiler-v2/tools/testdiff/src/main.rs:252]: Agree about not creating macro for this.

File: third_party/move/move-compiler-v2/tools/testdiff/src/main.rs (hunk 1 above)
Comment (brmataptos) [reply third_party/move/move-compiler-v2/tools/testdiff/src/main.rs:56]: Yes, here we are removing it because the attribute tests are under parser.

File: third_party/move/move-compiler-v2/tools/testdiff/src/main.rs (hunk 3 above)
Comment (brmataptos) [reply third_party/move/move-compiler-v2/tools/testdiff/src/main.rs:207]: I first tried using those methods but seemed to see inconsistency with multiple-levels of extensions (`foo.exp1`.with_extension("exp2") versus `foo.exp`.set_extension("exp2")).  I may just have been confused, but when I just tried again I ran into other problems:

`extension` is only useable on a `Path`.  Once we grab the filename, it's an `osstr` and those methods don't work.   So I'm sticking with the current code, which is a little ugly, but at least it's isolated ugliness and does what I want.

**Review by vineethk — APPROVED** [review]
Approving, assuming the `Cargo.lock` changes will be reverted.

File: third_party/move/move-compiler-v2/tools/testdiff/src/main.rs (hunk 5 above)
Comment (brmataptos) [reply third_party/move/move-compiler-v2/tools/testdiff/src/main.rs:19]: done
